src/libraries/Common/src/System/Security/Cryptography/SlhDsa.cs

```diff
@@ -760,35 +947,51 @@ public static SlhDsa ImportSubjectPublicKeyInfo(ReadOnlySpan<byte> source)
         ///   </para>
         ///   <para>-or-</para>
         ///   <para>
+        ///     <paramref name="source" /> contains trailing data after the ASN.1 structure.
+        ///   </para>
+        ///   <para>-or-</para>
+        ///   <para>
         ///     The algorithm-specific import failed.
         ///   </para>
         /// </exception>
+        /// <exception cref="PlatformNotSupportedException">
+        ///   The platform does not support SLH-DSA. Callers can use the <see cref="IsSupported" /> property
+        ///   to determine if the platform supports SLH-DSA.
+        /// </exception>
         public static SlhDsa ImportPkcs8PrivateKey(ReadOnlySpan<byte> source)
         {
+            ThrowIfInvalidLength(source);
             ThrowIfNotSupported();
 
-            unsafe
-            {
-                fixed (byte* pointer = source)
+            KeyFormatHelper.ReadPkcs8(
+                s_knownOids,
+                source,
+                (ReadOnlyMemory<byte> key, in AlgorithmIdentifierAsn algId, out SlhDsa ret) =>
                 {
-                    using (PointerMemoryManager<byte> manager = new(pointer, source.Length))
-                    {
-                        AsnValueReader reader = new AsnValueReader(source, AsnEncodingRules.DER);
-                        PrivateKeyInfoAsn.Decode(ref reader, manager.Memory, out PrivateKeyInfoAsn pki);
+                    SlhDsaAlgorithm info = GetAlgorithmIdentifier(in algId);
+                    ReadOnlySpan<byte> privateKey = key.Span;
 
-                        SlhDsaAlgorithm info = SlhDsaAlgorithm.GetAlgorithmFromOid(pki.PrivateKeyAlgorithm.Algorithm);
+                    if (privateKey.Length != info.SecretKeySizeInBytes)
+                    {
+                        throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                    }
 
-                        if (pki.PrivateKeyAlgorithm.Parameters.HasValue)
-                        {
-                            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
-                            pki.PrivateKeyAlgorithm.Encode(writer);
-                            throw Helpers.CreateAlgorithmUnknownException(writer);
-                        }
+                    ret = ImportSlhDsaSecretKey(info, key.Span);
+                },
+                out int read,
+                out SlhDsa kem);
+            Debug.Assert(read == source.Length);
```

Comment:
I'd also have a blank line before this assert, because I almost always have a blank line (or close-curly) after a multi-line statement.  I find it easier on the eyes.

Look at how well I'm being quiet about that preference, though.